content/api/servo-assembly/arduino-commands.md

```diff
@@ -0,0 +1,209 @@
+---
+title: Custom Arduino Commands
+---
+
+At the core of the servo assembly is an [Arduino Uno](https://store.arduino.cc/arduino-uno-rev3).
+By default the Arduino runs firmware created by SourceBots providing simple
+access to the attached servo shield, compatible ultrasound sensors and to the
+GPIO pins directly.
+
+You can also extend the firmware on the Arduino to support more features, for
+example if you have other types of sensor which require higher precision timing
+than can be achieved through the Python API.
+
+## Requirements
+
+As your new firmware will still need to be able to talk to the Python code
+running on the Raspberry Pi, you'll need to base it on SourceBots default
+firmware. You can get a copy of the source code from its GitHub repo at
+<https://github.com/sourcebots/servo-firmware>. Note that you will also need the
+Adafruit PWM Servo Driver Library which is included in that repo as a submodule.
+
+We recommend using [Arduino Studio](https://www.arduino.cc/en/Main/Software) to
+edit your firmware as it provides direct support for building the code into the
+firmware binary and downloading it onto the Arduino.
+
+## Adding a command
+
+In order for the code running on your Pi to be able to interact with the
+functionality you're adding to the Arduino, you'll need to define a new command.
+
+Adding a command requires two changes. First you must define the function which
+will implement the operations of that command and then you register that command
+under the name you'll use to call it from the Python code.
+
+### Implementing the command
+
+You'll need to define a function to implement your command. The function should
+be placed in the `firmware.cpp` file just above the definition of the `commands`
+array (likely about line 205).
```

Comment:
like -> line?

I feel like line numbers are quite a volatile way to refer to parts of the code, since changes to the file can result in many of the line numbers changing. IMO better solutions are to either copy the line verbatim here so users can `ctrl+f` for it, or just to say the rough location qualitatively (e.g. "about halfway through the file") - both of these methods are less likely to break when the file is modified.

Reply:
I'd hope that this wouldn't change very much, also hence the "about" on the line number. Given that we do include the definition of the commands array below I don't want to repeat it here.

Do you think we need to add a comment like "see below" for this, or can we assume people will be able to read the whole of the documentation to get a full understanding?

Reply:
I think we could just remove the remark - I think it should be easy enough to find `commands` in the source file.

Reply:
I think it's useful to have the line number; I think it's clear enough that it's approximate.